runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,156 @@
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import io.aklivity.zilla.runtime.engine.config.NamespaceConfig;
+
+public class FileWatcherTask extends WatcherTask
+{
+    private final Map<WatchKey, WatchedConfig> watchedConfigs;
+    private final MessageDigest md5;
+    private final WatchService watchService;
+
+    public FileWatcherTask(
+        BiFunction<URL, String, NamespaceConfig> changeListener)
+    {
+        super(changeListener);
+        this.watchedConfigs = new IdentityHashMap<>();
+        MessageDigest md5 = null;
+        WatchService watchService = null;
+
+        try
+        {
+            md5 = MessageDigest.getInstance("MD5");
+            watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch (NoSuchAlgorithmException | IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+
+        this.md5 = md5;
+        this.watchService = watchService;
+
+    }
+
+    @Override
+    public Void call()
+    {
+        while (true)
+        {
+            try
+            {
+                final WatchKey key = watchService.take();
+
+                WatchedConfig watchedConfig = watchedConfigs.get(key);
+
+                if (watchedConfig != null && watchedConfig.isWatchedKey(key))
+                {
+                    // Even if no reconfigure needed, recalculation is necessary, since symlinks might have changed.
+                    watchedConfig.keys().forEach(watchedConfigs::remove);
+                    watchedConfig.unregister();
+                    watchedConfig.register();
+                    watchedConfig.keys().forEach(k -> watchedConfigs.put(k, watchedConfig));
+                    String newConfigText = readConfigText(watchedConfig.getURL());
+                    byte[] newConfigHash = computeHash(newConfigText);
+                    if (watchedConfig.isReconfigureNeeded(newConfigHash))
+                    {
+                        watchedConfig.setConfigHash(newConfigHash);
+                        changeListener.apply(watchedConfig.getURL(), newConfigText);
+                    }
+                }
+            }
+            catch (InterruptedException | ClosedWatchServiceException ex)
+            {
+                break;
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public NamespaceConfig watch(
+        URL configURL)
+    {
+        WatchedConfig watchedConfig = new WatchedConfig(configURL, watchService);
+        watchedConfig.register();
+        watchedConfig.keys().forEach(k -> watchedConfigs.put(k, watchedConfig));
+        String configText = readConfigText(configURL);
+        watchedConfig.setConfigHash(computeHash(configText));
+        return changeListener.apply(configURL, configText);
+    }
+
+    @Override
+    public void doInitialConfiguration(
```

Comment:
Is this separate concept of `doInitialConfiguration` still needed?
Perhaps the caller can take this responsibility instead?